### PR TITLE
Fix further hash confusion between raw and merkle hash

### PR DESCRIPTION
**Pull request by Martin2112**
The LeafData must have the raw SHA256 hash as we use it for deduping. The sequenced data needs both raw and merkle hashes as it must both link to the LeafData and be queried by Merkle Hash by clients. I've renamed database columns to make it clearer which is which.

### CODE REVIEW

File: storage/mysql/log_storage.go [hunk 1]
```diff
@@ -18,32 +18,34 @@ import (
 
 const getTreePropertiesSQL string = "SELECT AllowsDuplicateLeaves FROM Trees WHERE TreeId=?"
 const getTreeParametersSQL string = "SELECT ReadOnlyRequests From TreeControl WHERE TreeID=?"
-const selectQueuedLeavesSQL string = `SELECT LeafHash,Payload
+const selectQueuedLeavesSQL string = `SELECT LeafRawHash,Payload
```
Comment (AlCutter) [inline storage/mysql/log_storage.go]: `DataHash`?

File: storage/mysql/log_storage.go [hunk 2]
```diff
@@ -18,32 +18,34 @@ import (
...
-     VALUES(?,?,?,?)`
-const insertSequencedLeafSQL string = `INSERT INTO SequencedLeafData(TreeId,LeafHash,SequenceNumber)
+		 ORDER BY QueueTimestamp DESC,LeafRawHash ASC LIMIT ?`
+const insertUnsequencedLeafSQL string = `INSERT INTO LeafData(TreeId,LeafRawHash,TheData)
+		 VALUES(?,?,?) ON DUPLICATE KEY UPDATE LeafRawHash=LeafRawHash`
+const insertUnsequencedLeafSQLNoDuplicates string = `INSERT INTO LeafData(TreeId,LeafRawHash,TheData)
 		 VALUES(?,?,?)`
+const insertUnsequencedEntrySQL string = `INSERT INTO Unsequenced(TreeId,LeafRawHash,MessageId,Payload)
+     VALUES(?,?,?,?)`
+const insertSequencedLeafSQL string = `INSERT INTO SequencedLeafData(TreeId,LeafRawHash,LeafHash,SequenceNumber)
```
Comment (AlCutter) [inline storage/mysql/log_storage.go]: Would renaming `LeafHash` to `MerkleLeafHash` help with clarity too?

File: storage/mysql/log_storage.go [hunk 3]
```diff
@@ -274,12 +276,23 @@ func (t *logTX) QueueLeaves(leaves []trillian.LogLeaf) error {
...
+		insertSQL = insertUnsequencedLeafSQLNoDuplicates
+	}
+
 	for _, leaf := range leaves {
 		// Create the unsequenced leaf data entry. We don't use INSERT IGNORE because this
 		// can suppress errors unrelated to key collisions. We don't use REPLACE because
 		// if there's ever a hash collision it will do the wrong thing and it also
 		// causes a DELETE / INSERT, which is undesirable.
-		_, err := t.tx.Exec(insertUnsequencedLeafSQL, t.ls.logID.TreeID,
+		_, err := t.tx.Exec(insertSQL, t.ls.logID.TreeID,
```
Comment (AlCutter) [inline storage/mysql/log_storage.go]: Might be a thing for another PR, but should this be using a prepared stmt?

File: storage/mysql/log_storage.go [hunk 4]
```diff
@@ -474,7 +487,11 @@ func (t *logTX) UpdateSequencedLeaves(leaves []trillian.LogLeaf) error {
 			return errors.New("Sequenced leaf has incorrect hash size")
 		}
 
-		_, err := t.tx.Exec(insertSequencedLeafSQL, t.ls.logID.TreeID, []byte(leaf.LeafHash),
+		// Recompute the raw leaf hash, could be passed around in the leaf structure but
```
Comment (AlCutter) [inline storage/mysql/log_storage.go]: Would it make sense to add an extra field to hold it?

File: storage/mysql/log_storage.go (hunk 1 above)
Comment (Martin2112) [inline storage/mysql/log_storage.go]: Not sure which is best. We use "leaf" a lot in the code but not sure we're always consistent.

File: storage/mysql/log_storage.go (hunk 2 above)
Comment (Martin2112) [inline storage/mysql/log_storage.go]: It might. I can do that.

File: storage/mysql/log_storage.go (hunk 3 above)
Comment (Martin2112) [inline storage/mysql/log_storage.go]: If we're going to do it for potential efficiency gains we should probably look at the other transaction level prepared statements. I don't think that belongs in this PR.

File: storage/mysql/log_storage.go (hunk 1 above)
Comment (AlCutter) [inline storage/mysql/log_storage.go]: or `TheDataHash`, just to make it clear that it's a hash of `TheData` (similar relative naming could happen in the `LeafData` struct.

File: storage/mysql/log_storage.go (hunk 4 above)
Comment (Martin2112) [inline storage/mysql/log_storage.go]: It does but we're again getting into slightly weird territory because the hash is part of the Leaf type, not LogLeaf.

Comment (Martin2112) [inline storage/mysql/log_storage.go]: Working on this now. I'm going to take the hit now and rename everything in Leaf / Logleaf end to end so it's always consistent.

File: storage/mysql/log_storage.go (hunk 2 above)
Comment (AlCutter) [inline storage/mysql/log_storage.go]: I think `Leaf`is only used by the Log...
Anyway, don't worry about it for now as I think we're going to have to move those around a bit.
Pop a TODO in?

**Comment by Martin2112** [comment]
OK I think this is all the renaming except for splitting off LeafValueHash, which can be done separately.

File: storage/mysql/log_storage.go (hunk 4 above)
Comment (Martin2112) [inline storage/mysql/log_storage.go]: TODO added.

**Comment by Martin2112** [comment]
PTAL.
